### ISSUE TITLE
Early stopping on test set + pretraining on ImageNet

### ISSUE THREAD

**Issue by TheOnlyJohnDoe** [xiaocangshengGR/DDGR#6]
Dear Authors,
First of all, thank you very much for sharing your code; it is great and very important to do so.

Unfortunately I have just realized that you are ___early stopping on the testing set___ which should not be happening as it is ___leaking test data.___
Furthermore, since you are using a CNN pretrained on ImageNet (--model_name alexnetCI_pretrained), the impacts of this  ___unintentional mistake is much more significant on the final accuracy scores which are greatly increased.___

Please see the explanations below

I initially wanted to reproduce your results on CIFAR100 using the two scripts provided in "run_cifar100.py".
I managed to obtain results comparable to the ones in Table 1 of the paper which is great, see the final logs below, as well as the manually calculated average accuracy.

--CI_task_count 10
------------------------------------------------------------------------------------------------------------
Testing on  5  task datasets
Testing_model: 0	Accuracy: 59.02000000	forgetting: 0.00000000
Testing_model: 1	Accuracy: 59.01000000	forgetting: 16.30000000
Testing_model: 2	Accuracy: 60.35333333	forgetting: 18.12000000
Testing_model: 3	Accuracy: 56.60000000	forgetting: 24.35500000
Testing_model: 4	Accuracy: 58.44800000	forgetting: 23.07600000

Average acc = 58.69 --- (59.02000000 + 59.01000000 + 60.35333333 + 56.60000000 + 58.44800000) / 5

------------------------------------------------------------------------------------------------------------

--CI_task_count 5
------------------------------------------------------------------------------------------------------------
Testing on  5  task datasets
Testing_model: 0	Accuracy: 58.86000000	forgetting: 0.00000000
Testing_model: 1	Accuracy: 46.38000000	forgetting: 18.05000000
Testing_model: 2	Accuracy: 53.92000000	forgetting: 12.70000000
Testing_model: 3	Accuracy: 51.78000000	forgetting: 17.23500000
Testing_model: 4	Accuracy: 50.81600000	forgetting: 19.77600000

Average acc = 52.35 --- (58.86000000 + 46.38000000 + 53.92000000 + 51.78000000 + 50.81600000) / 5

------------------------------------------------------------------------------------------------------------
By the way, those scripts were not training on the entire CIFAR100 dataset, but only on 5 tasks, and I did not realize this when reading the paper since it is ___not common at all___, and only mentioned in the supplementary material. I am writing this down here for other people to see, since this already means that the results cannot be compared with other works that run the standard benchmarks with 6 or 11 task ( For more details see Table A.1 in the supplementary material page 13.)

Now since the previous scores used a model pretrained on ImageNet, I wanted to try the same benchmark ___without the pretrained alexnet___, (--model_name alexnetCI), and got much lower results (see logs below). 

--CI_task_count 10
------------------------------------------------------------------------------------------------------------
Testing on  5  task datasets
Testing_model: 0	Accuracy: 2.72000000	 forgetting: 0.00000000
Testing_model: 1	Accuracy: 15.86000000	forgetting: 0.30000000
Testing_model: 2	Accuracy: 19.45333333	forgetting: -1.88000000
Testing_model: 3	Accuracy: 19.03500000	forgetting: -1.35500000
Testing_model: 4	Accuracy: 18.95200000	forgetting: -1.08800000

Average acc = 15.20 --- (2.72000000 + 15.86000000 + 19.45333333 + 19.03500000 + 18.95200000) / 5

------------------------------------------------------------------------------------------------------------

--CI_task_count 5
------------------------------------------------------------------------------------------------------------
Testing on  5  task datasets
Testing_model: 0	Accuracy: 2.72000000	 forgetting: 0.00000000
Testing_model: 1	Accuracy: 6.40000000 forgetting: -0.04000000
Testing_model: 2	Accuracy: 7.36666667	 forgetting: -0.02666667
Testing_model: 3	Accuracy: 7.91500000	 forgetting: 0.06500000
Testing_model: 4	Accuracy: 8.41600000	 forgetting: 0.06800000

Average acc = 6.56 --- (2.72000000 + 6.40000000 + 7.36666667	 + 7.91500000 + 8.41600000) / 5

------------------------------------------------------------------------------------------------------------
Such a difference in performance, especially at task 1, which only reaches 2.72% when the classifier is trained on the first task is very intriguing.

After looking through the code, I realized that you were ___early stopping on validation set identical to the test set___ using the function "def set_lr(optimizer, lr, count):" in train_DDGR.py and the "val_beat_counts" variable which stops the training using the "continue_training" variable, and  the function "def traminate_protocol(since, best_acc):". 

Indeed, the images in val and test folders are identical (by visual inspection), and this also be confirmed by adding the following lines of code at line 51 in "main_DDGR.py", to print a portion of them. 
```
    print("those are the dsets")
    for key, val in dsets.items():
        if key != "train": 
            print(key, val)
            for i in range(10):
                print(val.samples[i])

    import sys
    sys.exit(0)
    
    dset_sizes = {x: len(dsets[x]) for x in ['train', 'val']}
    dset_classes = dsets['train'].classes
```

Alternatively to make sure of it, one may also visually inspect the images used in the actual validation loop by following the steps below :
replace "shuffle=True" with "shuffle=(x=='train')" on lines 222, 231 and 236 of main_DDGR.py
then add the following lines of code between line 77 and 78 of train_DDGR.py

```
                # This is added (below)
                # ------------------------------------------------------------------------------------
                if phase == "val":
                    import torchvision
                    import torchvision.transforms as transforms

                    def reverse_transform(
                        x,
                        mean=(0.5070751592371323, 0.48654887331495095, 0.4409178433670343), 
                        std=(0.2673342858792401, 0.2564384629170883, 0.27615047132568404)
                    ):
                    
                        inv_trans = transforms.Compose([
                            transforms.Normalize(
                                mean=[-m / s for m, s in zip(mean, std)],
                                std=[1 / s for s in std]
                            ),
                        ])
                        x = inv_trans(x)
                        return x

                    k_kept=16
                    x_kept = reverse_transform(inputs[:k_kept])
                    grid = torchvision.utils.make_grid(x_kept, nrow=int(k_kept**0.5)).cpu()
                    torchvision.utils.save_image(grid, "val_img.png")
                    sys.exit(0)
                # ------------------------------------------------------------------------------------
                # This is added (above)
```
Then by running an experiment from scratch (do not forget to delete "results" folder or rename it to avoid resuming), this will save an image called val_img.png and exit the code.

Now by running the code below in a jupyter notebook, which loads the testing images for CIFAR100 dataset, it is possible to recognise the first images from the test set from class 0 (apple), 8(bike) or 43(lion) that are in the validation set -  val_img.png. 

```
import torch
import torchvision
import torchvision.transforms as transforms
import matplotlib.pyplot as plt

transform = transforms.Compose([transforms.ToTensor()])
trainset = torchvision.datasets.CIFAR100(root='./data', train=False, download=True, transform=transform)

for label_choice in [0, 8, 43]:
    x = [img for (img, label) in trainset if label == label_choice]
    x = torch.stack(x)
    k_kept = 10 # enough to see the first test images from each class since it is not shuffled
    x = x[:k_kept]
    grid = torchvision.utils.make_grid(x, nrow=k_kept)

    image_np = grid.permute(1, 2, 0).numpy()
    plt.figure(figsize=(10, 10))
    plt.imshow(image_np)
    plt.gca().axis('off')
    plt.tight_layout()
    plt.show()
```

Finally you can see the images obtained following this protocol. If you were to continue you would find that all validation images are in fact from the test set.

![Image](https://github.com/user-attachments/assets/0a4f0258-d9fd-48f1-bdb8-b15857e08445)

Unfortunately, this means that the results from Table 1 in this paper ___cannot be compared with other works since the benchmark used is erroneous / uncommon.___ (however you wanna call it). Indeed, the scores in all the tables can only be compared with scores that ran the exact same benchmark that is :
- early stopping on the test set,
- imagenet pretraining for the classifier,
- only training and testing on 5 tasks (for CIFAR100 and Imagenet), which ___does not amount to the entire dataset___

Looking forward to your answers on this, and please do correct me if you think that I am wrong,
Best Regards
